vmm/src/cpu.rs

```diff
@@ -422,42 +422,25 @@ impl Vcpu {
     #[cfg(target_arch = "aarch64")]
     pub fn init(&self, vm: &Arc<dyn hypervisor::Vm>) -> Result<()> {
         use std::arch::is_aarch64_feature_detected;
-        let mut kvi: kvm_bindings::kvm_vcpu_init = kvm_bindings::kvm_vcpu_init::default();
         #[allow(clippy::nonminimal_bool)]
         let sve_supported =
             is_aarch64_feature_detected!("sve") || is_aarch64_feature_detected!("sve2");
+        let mut kvi = self.vcpu.create_vcpu_init();
+
         // This reads back the kernel's preferred target type.
         vm.get_preferred_target(&mut kvi)
             .map_err(Error::VcpuArmPreferredTarget)?;
-        // We already checked that the capability is supported.
-        kvi.features[0] |= 1 << kvm_bindings::KVM_ARM_VCPU_PSCI_0_2;
-        if vm
-            .as_any()
-            .downcast_ref::<hypervisor::kvm::KvmVm>()
-            .unwrap()
-            .check_extension(Cap::ArmPmuV3)
-        {
-            kvi.features[0] |= 1 << kvm_bindings::KVM_ARM_VCPU_PMU_V3;
-        }
 
-        if sve_supported
-            && vm
-                .as_any()
-                .downcast_ref::<hypervisor::kvm::KvmVm>()
-                .unwrap()
-                .check_extension(Cap::ArmSve)
-        {
-            kvi.features[0] |= 1 << kvm_bindings::KVM_ARM_VCPU_SVE;
-        }
+        self.vcpu
+            .vcpu_set_processor_features(vm, &mut kvi, self.id)
+            .map_err(Error::VcpuSetProcessorFeatures)?;
 
-        // Non-boot cpus are powered off initially.
-        if self.id > 0 {
-            kvi.features[0] |= 1 << kvm_bindings::KVM_ARM_VCPU_POWER_OFF;
-        }
         self.vcpu.vcpu_init(&kvi).map_err(Error::VcpuArmInit)?;
+
         if sve_supported {
+            let finalized_features = self.vcpu.vcpu_get_finalized_features();
             self.vcpu
-                .vcpu_finalize(kvm_bindings::KVM_ARM_VCPU_SVE as i32)
+                .vcpu_finalize(finalized_features)
```

Comment:
These changes are refactoring, and should be a part of the refactoring commit.

Reply:
Done